### PR TITLE
A batch of locale handling patches

### PR DESCRIPTION
**Pull request by khwilliamson**

### CODE REVIEW

File: locale.c [hunk 1]
```diff
@@ -1730,6 +1725,16 @@ S_calculate_LC_ALL_string(pTHX_ const char ** category_locales_list)
      * categories, adding new ones as they show up on obscure platforms.
      */
 
+    const char * my_category_locales_list[LOCALE_CATEGORIES_COUNT_];
+    const char ** locales_list = category_locales_list;
+    if (locales_list == NULL) {
+        locales_list = (const char **) &my_category_locales_list;
```
Comment (tonycoz) [inline locale.c]: This should just be:
```
        locales_list = my_category_locales_list;
```
An array of T  (T = `const char *` here) decays to a pointer to T in most cases in C (`const char **` in this case).

The `&` on the array name is one of the few exceptions\*, which would make `&my_category_locales_list` a pointer to an array of LOCALE_CATEGORIES_COUNT_ pointers to const char *, which I suspect led you to adding the cast.

\* the other one I can think of is sizeof().

File: hints/darwin.sh [hunk 2]
```diff
@@ -357,8 +357,7 @@ EOM
         esac
     fi
 
-    # The OS is buggy with respect to this.
```
Comment (tonycoz) [inline hints/darwin.sh:360]: commit message, perhaps:

Make sure *it* is executed while a mutex locks out other accesses.

Perhaps also "has *a* race" in the subject

File: locale.c [hunk 3]
```diff
@@ -768,6 +768,200 @@ Perl_locale_panic(const char * msg,
...
+     * caller needs to arrange for each to be freed.
+     *
+     */
+
+    DEBUG_Lv(PerlIO_printf(Perl_debug_log,
+                           "Entering parse_LC_ALL_string; called from %"    \
+                           LINE_Tf "\nnew='%s'\n", caller_line, string));
+
+    const char separator[] = ";";
+    const Size_t separator_len = 1;
```
Comment (tonycoz) [inline locale.c:815]: Are these ever not ";"? Or not a single character?

I do see this has changed from the code it replaces which uses character checks for ';'.

File: locale.c [hunk 4]
```diff
@@ -1401,114 +1595,48 @@ S_setlocale_from_aggregate_LC_ALL(pTHX_ const char * locale, const line_t line)
...
-            p++;
-        }
+      case full_array:
+        break;
+    }
 
-        /* Find the index of the category name in our lists */
-        for (PERL_UINT_FAST8_T i = 0; i < LC_ALL_INDEX_; i++) {
+    /* Change each category to the value returned for it */
+    for (unsigned int i = 0; i < LC_ALL_INDEX_; i++) {
```
Comment (tonycoz) [inline locale.c]: I do prefer the more general types here.  While it's not relevant here (LC_ALL_INDEX_ is a small integer), in some other cases CodeQL complains about using the possibly-small-PERL_UINT_FAST8_T variable as a loop index when the loop limit is a bigger type, for example:
```
regexec.c:4952
        for (i = 0; i < folds_to_count
                      + UNLIKELY(multi_fold_from != 0); i++)
Comparison between i of type PERL_UINT_FAST8_T and ... + ... of wider type unsigned long.
```
(this is as https://github.com/tonycoz/perl5/security/code-scanning/27 which I'm not sure you can see)

File: locale.c (hunk 3 above)
Comment (khwilliamson) [reply locale.c:815]: A future commit will take advantage of this being a variable, and with different potential values.

I have not seen any instances of more than a single character, but it seemed less brittle to allow a string for any such platforms, and with no extra logic needed.

File: locale.c (hunk 4 above)
Comment (khwilliamson) [reply locale.c]: A later commit will change to declare `i` to be an enum; I'm unsure if that is a good idea, but we can defer discussion until then.

Comment (tonycoz) [reply locale.c]: You can't increment an enum type in C++:
```
tony@venus:.../perl/git$ g++ -oenuminc enuminc.cpp 
enuminc.cpp: In function ‘int main()’:
enuminc.cpp:9:33: error: no match for ‘operator++’ (operand type is ‘foo’)
    9 |     for (foo i = a; i < nomore; ++i)
      |                                 ^~~
tony@venus:.../perl/git$ cat enuminc.cpp
#include <iostream>

enum foo {
    a, b, nomore
};

int
main() {
    for (foo i = a; i < nomore; ++i)
        std::cout << i << '\n';
}
```
unless you provide an overloaded operator:
```
tony@venus:.../perl/git$ g++ -oenuminc enuminc.cpp 
tony@venus:.../perl/git$ cat enuminc.cpp
#include <iostream>

enum foo {
    a, b, nomore
};

foo operator++(foo &x) {
    x = static_cast<foo>(x+1);
    return x;
}

int
main() {
    for (foo i = a; i < nomore; ++i)
        std::cout << i << '\n';
}
tony@venus:.../perl/git$ ./enuminc
0
1
```

File: hints/darwin.sh (hunk 2 above)
Comment (khwilliamson) [reply hints/darwin.sh:360]: done

File: locale.c (hunk 1 above)
Comment (khwilliamson) [reply locale.c]: done

File: locale.c (hunk 4 above)
Comment (khwilliamson) [reply locale.c]: Right, but I think it is better to wait to discuss this until we get to a commit that actually might use this

**Review by tonycoz — APPROVED** [review]

**Pull request merged**